### PR TITLE
luci-statistics: rework cpu and memory graph definitions

### PR DESCRIPTION
**Pull request by hippi-viking**
Simplify CPU and memory definitions, make them FreeBSD compatible

Signed-off-by: Istvan S Sz <hippi-viking@protonmail.org>

### CODE REVIEW

**Comment by hnyman** [comment]
What do you mean by "FreeBSD compatible" ?  
LuCI is just for OpenWrt, which is not FreeBSD.

**Comment by hippi-viking** [comment]
I am using luci-statistics to collect and display telemetry of very different devices - some running Linux, some running FreeBSD. CPU and memory graphs from FreeBSD devices are not parsed by luci-statistics without this patch because of different nomenclature used by the OS.

**Comment by systemcrash** [comment]
Is openwrt still parsed after this patch? If so I think it would be okay to merge.

**Comment by systemcrash** [comment]
Welp, in my naïve view, this looks acceptable. 

I like the code reduction. Just needs verifying on openwrt :)

**Comment by hnyman** [comment]
@jow- 
Any opinion on this?

**Comment by jow-** [comment]
If it is confirmed to still work normally with/for OpenWrt then I'm fine with merging it.

**Comment by raenye** [comment]
I tested this.
Works fine but the order of fields has changed (probably due to the use of a dictionary or however hash-maps are called in JS)
But I like the code reduction too.

Before:
image
image

After:
image
image

**Comment by systemcrash** [comment]
OK, I'm satisfied with that.

**Comment by hnyman** [comment]
Based on the screenshots, the display order changes are really bad.

The current orders are based on "system level interrupts" / "used" at bottom, and then having more userland/optional/cached items gradually , and "free" at the top.

The proposed change completely destroys that logic, as it appears to show them in alphabetically sorted :-(    E.g. that memory graph is crazy, as free mem is in the middle.

**Comment by systemcrash** [comment]
Does not bother me one bit. Everything is colour coded. Alphabetical order is logical for even a school child.

Have you a suggestion to fix ordering in JS?

**Comment by hnyman** [comment]
> Have you a suggestion to fix ordering in JS?

The current status ;-)

File: applications/luci-app-statistics/htdocs/luci-static/resources/statistics/rrdtool/definitions/memory.js [hunk 1]
```diff
@@ -10,89 +10,63 @@ return baseclass.extend({
...
-						title: "Cached"
-					},
-					memory_used: {
-						color: "ff0000",
-						title: "Used"
-					},
-					memory_free: {
-						color: "00ff00",
-						title: "Free"
+			return matching.length ? { [dtype]: matching } : null;
```
Comment (raenye) [inline applications/luci-app-statistics/htdocs/luci-static/resources/statistics/rrdtool/definitions/memory.js:21]: I think this needs to be changed to a list, and then it'll maintain the order.

Comment (raenye) [reply applications/luci-app-statistics/htdocs/luci-static/resources/statistics/rrdtool/definitions/memory.js:21]: Here's a working solution (still [quadratic](https://accidentallyquadratic.tumblr.com/)).

```js
function find_instances(dtype, wanted) {
  var di = graph.dataInstances(host, plugin, plugin_instance, dtype);
  var matching = wanted.filter(w => di.includes(w));
  return matching.length ? { [dtype]: matching } : null;
}
````
Should be placed in both `cpu.js` and `memory.js`.

Comment (systemcrash) [reply applications/luci-app-statistics/htdocs/luci-static/resources/statistics/rrdtool/definitions/memory.js:21]: @hippi-viking could you implement the fix?

Comment (hippi-viking) [reply applications/luci-app-statistics/htdocs/luci-static/resources/statistics/rrdtool/definitions/memory.js:21]: @raenye thanks for the solution proposal.
@systemcrash I'll have a go at it, but TBH I'm more of a backend-guy not really fluent in javascript. (And TBH it's good enough for my use as well the way it is, but when I'll have some time I'll definitely try to make it more aesthetically pleaseing.)

Comment (raenye) [reply applications/luci-app-statistics/htdocs/luci-static/resources/statistics/rrdtool/definitions/memory.js:21]: (while you're at it, you should squash the commits and write a full name in the signoff line)

**Comment by feckert** [comment]
@systemcrash Please write the next time why you decided to close this.

**Comment by systemcrash** [comment]
Sure: @hnyman (I think) already committed changes which break this update without further changes to the PR and the original committer does not show signs of updating it, where the original intention was incidental ("add this if it's useful"). We can reopen this @hippi-viking should you decide to update this.

**Pull request closed without merging**